### PR TITLE
TSQL: extend security statement

### PR DESCRIPTION
**Pull request by scrappyCoco**
`class_type_for_grant` has been added for `GRANT` and `grant_permission` has been extended.

### CODE REVIEW

**Review by KvanTTT — APPROVED** [review]

File: sql/tsql/TSqlParser.g4 [hunk 1]
```diff
@@ -626,6 +626,69 @@ class_type_for_parallel_dw
...
+    | CONTRACT
+    | CREDENTIAL
+    | CRYPTOGRAPHIC PROVIDER
+    | DATABASE
+    | DATABASE AUDIT SPECIFICATION
+    | DATABASE ENCRYPTION KEY
+    | DATABASE EVENT SESSION
+    | DATABASE SCOPED CONFIGURATION
+    | DATABASE SCOPED CREDENTIAL
+    | DATABASE SCOPED RESOURCE GOVERNOR
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:652]: What do you think about simplifying by the following way? Here and in other places.

```suggestion
    | DATABASE
        ( AUDIT SPECIFICATION
        | ENCRYPTION KEY
        | EVENT SESSION
        | SCOPED ( CONFIGURATION
                 | CREDENTIAL
                 | RESOURCE GOVERNOR )
        )
```

It's good both for performance and clarity.

Comment (scrappyCoco) [reply sql/tsql/TSqlParser.g4]: Thanks for your suggestion! I have fixed it. Hope now is all right.

File: sql/tsql/TSqlParser.g4 [hunk 2]
```diff
@@ -626,6 +626,61 @@ class_type_for_parallel_dw
...
+    | CRYPTOGRAPHIC PROVIDER
+    | DATABASE ( AUDIT SPECIFICATION
+               | ENCRYPTION KEY
+               | EVENT SESSION
+               | SCOPED ( CONFIGURATION
+                        | CREDENTIAL
+                        | RESOURCE GOVERNOR )
+               )?
+    | ENDPOINT
+    | EVENT SESSION | ( NOTIFICATION (DATABASE | OBJECT | SERVER ) )
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:653]: Minor, but it should be on separated lines:

```suggestion
    | EVENT SESSION
    | ( NOTIFICATION (DATABASE | OBJECT | SERVER ) )
```

File: sql/tsql/TSqlParser.g4 [hunk 3]
```diff
@@ -2593,7 +2649,7 @@ security_statement
     // https://msdn.microsoft.com/en-us/library/ms188354.aspx
     : execute_clause ';'?
     // https://msdn.microsoft.com/en-us/library/ms187965.aspx
-    | GRANT (ALL PRIVILEGES? | grant_permission ('(' column_name_list ')')?) (ON on_id=table_name)? TO to_principal+=principal_id (',' to_principal+=principal_id)* (WITH GRANT OPTION)? (AS as_principal=principal_id)? ';'?
+    | GRANT (ALL PRIVILEGES? | grant_permission ('(' column_name_list ')')?) (ON (class=class_type_for_grant '::')? on_id=table_name)? TO to_principal+=principal_id (',' to_principal+=principal_id)* (WITH GRANT OPTION)? (AS as_principal=principal_id)? ';'?
```
Comment (studentmain) [inline sql/tsql/TSqlParser.g4:2652]: class -> class_ to avoid symbol conflict

**Pull request merged**